packages/web-components/fast-foundation/src/virtual-list/virtual-list.ts

```diff
@@ -0,0 +1,885 @@
+import {
+    attr,
+    DOM,
+    Notifier,
+    nullableNumberConverter,
+    Observable,
+    observable,
+    RepeatBehavior,
+    RepeatDirective,
+    Splice,
+    ViewTemplate,
+} from "@microsoft/fast-element";
+import { eventResize, eventScroll, Orientation } from "@microsoft/fast-web-utilities";
+import { FoundationElement } from "../foundation-element";
+import { IntersectionService } from "../utilities/intersection-service";
+import { IdleCallbackQueue } from "../utilities/idle-callback-queue";
+import type { ResizeObserverClassDefinition } from "../utilities/resize-observer";
+/**
+ * Defines when the component updates its position automatically.
+ *
+ * @public
+ */
+export type VirtualListAutoUpdateMode = "manual" | "viewport-resize" | "auto";
+
+/**
+ * Used to describe the position of an element within the list
+ *
+ * @public
+ */
+export interface SpanMap {
+    start: number;
+    end: number;
+    span: number;
+}
+
+/**
+ *  The VirtualList class
+ *
+ * @public
+ */
+export class VirtualList extends FoundationElement {
+    /**
+     *  Whether or not the display should virtualize
+     *
+     * @public
+     */
+    @attr({ attribute: "virtualize", mode: "boolean" })
+    public virtualize: boolean = true;
+    private virtualizeChanged(): void {
+        if (this.$fastController.isConnected) {
+            this.reset();
+        }
+    }
+
+    /**
+     * The HTML ID of the viewport element
+     *
+     * @public
+     * @remarks
+     * HTML Attribute: anchor
+     */
+    @attr({ attribute: "viewport" })
+    public viewport: string = "";
+    private viewportChanged(): void {
+        if (this.$fastController.isConnected) {
+            this.viewportElement = this.getViewport();
+            this.updateDimensions();
+        }
+    }
+
+    /**
+     * The span in pixels of each item along the virtualization axis
+     *
+     * @public
+     * @remarks
+     * HTML Attribute: item-span
+     */
+    @attr({ attribute: "item-span", converter: nullableNumberConverter })
+    public itemSpan: number = 50;
+    private itemSpanChanged(): void {
+        if (this.$fastController.isConnected) {
+            this.updateDimensions();
+        }
+    }
+
+    /**
+     * Defines an area in pixels on either end of the viewport where items outside the viewport
```

Comment:
Any reason this is measured in px instead of items? Items seems like a good way to think about it, but maybe some advanced scenarios don't work well that way. Either way, maybe adopt whatever naming convention per above. I think "size" seems most clear.

Reply:
so "viewport-buffer-span" or "viewport-buffer-size"?  Seems reasonable.  

I chose "pixels" over "items" because I was thinking of cases where authors will want to render a full-page scroll's worth of items to easily accommodate page up/down functionality (ie. it's easier when the item where focus is going to actually exists).